src/main/java/com/iota/iri/service/transactionpruning/async/AsyncTransactionPruner.java

```diff
@@ -0,0 +1,398 @@
+     * background.
+     *
+     * Note: This method is thread safe since we use a {@link ThreadIdentifier} to address the {@link Thread}. The
+     *       {@link ThreadUtils} take care of only launching exactly one {@link Thread} that is not terminated.
+     */
+    public void start() {
+        ThreadUtils.spawnThread(this::processJobsThread, cleanupThreadIdentifier);
+        ThreadUtils.spawnThread(this::persistThread, persisterThreadIdentifier);
+    }
+
+    /**
+     * Shuts down the background job by setting the corresponding shutdown flag.
+     */
+    public void shutdown() {
+        ThreadUtils.stopThread(cleanupThreadIdentifier);
+        ThreadUtils.stopThread(persisterThreadIdentifier);
+    }
+
+    /**
+     * This method contains the logic for the processing of the cleanup jobs, that gets executed in a separate
+     * {@link Thread}.
+     *
+     * It repeatedly calls {@link #processJobs()} until the TransactionPruner is shutting down.
+     */
+    private void processJobsThread() {
+        while(!Thread.interrupted()) {
+            try {
+                processJobs();
+            } catch(TransactionPruningException e) {
+                log.error("error while processing the transaction pruner jobs", e);
+            }
+
+            ThreadUtils.sleep(GARBAGE_COLLECTOR_RESCAN_INTERVAL);
+        }
+    }
+
+    /**
+     * This method contains the logic for persisting the pruner state, that gets executed in a separate {@link Thread}.
+     *
+     * It periodically checks the {@link #persistRequested} flag and triggers the writing of the state file until the
+     * {@link AsyncTransactionPruner} is shutting down.
+     */
+    private void persistThread() {
+        while(!Thread.interrupted()) {
+            try {
+                if (persistRequested) {
+                    saveStateNow();
+
+                    persistRequested = false;
+                }
+            } catch(TransactionPruningException e) {
+                log.error("could not persist transaction pruner state", e);
+            }
+
+            ThreadUtils.sleep(GARBAGE_COLLECTOR_PERSIST_INTERVAL);
+        }
+    }
+
+    /**
+     * This method creates a serialized version of the given job.
+     *
+     * @param job job that shall get serialized
+     * @return serialized representation of the job
+     */
+    private String serializeJobEntry(TransactionPrunerJob job) {
+        return job.getClass().getCanonicalName() + ";" + job.serialize();
+    }
+
+    /**
+     * Saves the state by serializing the jobs into a state file that is stored on the hard disk of the node.
+     *
+     * If no jobs are queued it removes the state file.
+     *
+     * @throws TransactionPruningException if anything goes wrong while persisting the state
+     */
+    private void saveStateNow() throws TransactionPruningException {
+        try {
+            AtomicInteger jobsPersisted = new AtomicInteger(0);
```

Comment:
because you can not modify local variables in a lambda - the only way to modify variables outside of the scope of lambdas is to use mutable objects like this.